packages/plugin-workspaces-tools/sources/commands/foreach.ts

```diff
@@ -0,0 +1,128 @@
+import { Configuration, IdentHash, PluginConfiguration, Project, Workspace } from '@berry/core';
+import { structUtils, Cache, DescriptorHash, StreamReport }                  from '@berry/core';
+import { Writable }                                                          from 'stream';
+import { cpus }                                                              from 'os';
+// @ts-ignore
+import pLimit                                                                from 'p-limit';
+
+import * as workspaceUtils                                                   from '../workspaceUtils';
+
+type ForeachOptions = {
+  args: Array<string>;
+  command: string;
+  cwd: string;
+  exclude: string[];
+  include: string[];
+  interlaced: boolean;
+  parallel: boolean;
+  prefixed: boolean;
+  stdout: Writable;
+  withDependencies: boolean;
+}
+
+// eslint-disable-next-line arca/no-default-export
+export default (clipanion: any, pluginConfiguration: PluginConfiguration) => clipanion
+
+    .command(`workspaces foreach <command> [...args] [-p,--parallel] [--with-dependencies] [-I,--interlaced] [-P,--prefixed] [-i,--include WORKSPACES...] [-x,--exclude WORKSPACES...]`)
+    .flags({proxyArguments: true})
+
+    .categorize(`Workspace-related commands`)
+    .describe(`run a command on all workspaces`)
+
+    .action(
+      async ({cwd, args, stdout, command, exclude, include, interlaced, parallel, withDependencies, prefixed, ...env}: ForeachOptions) => {
+        const configuration = await Configuration.find(cwd, pluginConfiguration);
+        const { project } = await Project.find(configuration, cwd);
+        const cache = await Cache.find(configuration);
+
+        const needsProcessing = new Map<IdentHash, Workspace>();
```

Comment:
The key should be a `LocatorHash`, not `IdentHash` (the current code makes it possible to have multiple workspaces with the same name but different versions. Even if that changes we might want to keep using the locators as workspace identifiers to stay in line with how it works for other packages).